include/beast/websocket/detail/stream_base.hpp

```diff
@@ -115,6 +131,287 @@ struct stream_base
     write_ping(DynamicBuffer& db, opcode op, ping_data const& data);
 };
 
+template<class _>
+void
+stream_base::open(role_type role)
+{
+    // VFALCO TODO analyze and remove dupe code in reset()
+    role_ = role;
+    failed_ = false;
+    rd_need_ = 0;
+    rd_cont_ = false;
+    wr_close_ = false;
+    wr_cont_ = false;
+    wr_block_ = nullptr;    // should be nullptr on close anyway
+    pong_data_ = nullptr;   // should be nullptr on close anyway
+}
+
+// Read fixed frame header
+// Requires at least 2 bytes
+//
+template<class DynamicBuffer>
+std::size_t
+stream_base::
+read_fh1(DynamicBuffer& db, close_code::value& code)
+{
+    using boost::asio::buffer;
+    using boost::asio::buffer_copy;
+    using boost::asio::buffer_size;
+    auto const err =
+        [&](close_code::value cv)
+        {
+            code = cv;
+            return 0;
+        };
+    std::uint8_t b[2];
+    assert(buffer_size(db.data()) >= sizeof(b));
+    db.consume(buffer_copy(buffer(b), db.data()));
+    std::size_t need;
+    rd_fh_.len = b[1] & 0x7f;
+    switch(rd_fh_.len)
+    {
+        case 126: need = 2; break;
+        case 127: need = 8; break;
+        default:
+            need = 0;
+    }
+    rd_fh_.mask = (b[1] & 0x80) != 0;
+    if(rd_fh_.mask)
+        need += 4;
+    rd_fh_.op   = static_cast<opcode>(b[0] & 0x0f);
+    rd_fh_.fin  = (b[0] & 0x80) != 0;
+    rd_fh_.rsv1 = (b[0] & 0x40) != 0;
+    rd_fh_.rsv2 = (b[0] & 0x20) != 0;
+    rd_fh_.rsv3 = (b[0] & 0x10) != 0;
+    switch(rd_fh_.op)
+    {
+    case opcode::binary:
+    case opcode::text:
+        if(rd_cont_)
+        {
+            // new data frame when continuation expected
+            return err(close_code::protocol_error);
+        }
+        if(rd_fh_.rsv1 || rd_fh_.rsv2 || rd_fh_.rsv3)
+        {
+            // reserved bits not cleared
+            return err(close_code::protocol_error);
+        }
+        break;
+
+    case opcode::cont:
+        if(! rd_cont_)
+        {
+            // continuation without an active message
+            return err(close_code::protocol_error);
+        }
+        if(rd_fh_.rsv1 || rd_fh_.rsv2 || rd_fh_.rsv3)
+        {
+            // reserved bits not cleared
+            return err(close_code::protocol_error);
+        }
+        break;
+
+    default:
+        if(is_reserved(rd_fh_.op))
+        {
+            // reserved opcode
+            return err(close_code::protocol_error);
+        }
+        if(! rd_fh_.fin)
+        {
+            // fragmented control message
+            return err(close_code::protocol_error);
+        }
+        if(rd_fh_.len > 125)
+        {
+            // invalid length for control message
+            return err(close_code::protocol_error);
+        }
+        if(rd_fh_.rsv1 || rd_fh_.rsv2 || rd_fh_.rsv3)
```

Comment:
Hmm the current version is a bit more clear. Also, this line will change with permessage-deflate (which uses some of the now-reserved bits).